src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java

```diff
@@ -192,44 +190,18 @@ public void deleteDirectoryEntriesFromCollection(Firestore firestore, String col
         DocumentSnapshot apply(Transaction xn) throws InterruptedException;
     }
 
-    // Returns null if not found - upper layers do any throwing
-    private FireStoreDirectoryEntry runTransactionWithRetry(Firestore firestore,
-                                                            LookupFunction lookupFunction,
-                                                            String transactionOp,
-                                                            String warnMessage) throws InterruptedException {
-        for (int i = 0; i < LOOKUP_RETRIES; i++) {
-            try {
-                ApiFuture<FireStoreDirectoryEntry> transaction =
-                    firestore.runTransaction(
-                        xn -> {
-                            DocumentSnapshot docSnap = lookupFunction.apply(xn);
-                            if (docSnap == null) {
-                                return null;
-                            }
-                            return docSnap.toObject(FireStoreDirectoryEntry.class);
-                        });
-
-                return fireStoreUtils.transactionGet(transactionOp, transaction);
-            } catch (Exception ex) {
-                if (FireStoreUtils.shouldRetry(ex)) {
-                    logger.warn("Retry-able error in firestore future get - " +
-                        warnMessage + " message: " + ex.getMessage());
-                } else {
-                    throw new FileSystemExecutionException(transactionOp + " execution exception", ex);
-                }
-            }
-
-            TimeUnit.SECONDS.sleep(1);
-        }
-        throw new FileSystemExecutionException(transactionOp + " failed - no more retries");
-    }
+    private static final int SLEEP_BASE_SECONDS = 1;
 
     // Returns null if not found - upper layers do any throwing
     public FireStoreDirectoryEntry retrieveById(
         Firestore firestore, String collectionId, String fileId) throws InterruptedException {
 
-        return runTransactionWithRetry(firestore, (xn) -> lookupByFileId(firestore, collectionId, fileId, xn),
+        DocumentSnapshot docSnap = fireStoreUtils.runTransactionWithRetry(firestore,
+            xn -> lookupByFileId(firestore, collectionId, fileId, xn),
             "retrieveById", " file id: " + fileId);
+
+        return Optional.ofNullable(docSnap).map(d -> docSnap.toObject(FireStoreDirectoryEntry.class))
+            .orElse(null);
```

Comment:
OK, I was going to say we shouldn't be doing this, but then I realized that there's a better solution. Instead of making an `Optional` here, `FirestoreUtils. runTransactionWithRetry()` could return `Optional<T>` instead of `<T>`.

And really, _this_ method should return `Optional<FireStoreDirectoryEntry>`. But that's probably something for another PR.

Reply:
shouldn't be doing this - making an `Optional` out of an object just to test `!= null`

Reply:
Sounds like a great idea! My first pass at implementing this didn't go so well, so I'd like to put this in a follow on PR where I can test this change out more.